### PR TITLE
Travis CI: Upgrade to production Py37 and run flake8

### PR DESCRIPTION
**Pull request by cclauss**
$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__

__E901,E999,F821,F822,F823__ are the "_showstopper_" [flake8](http://flake8.pycqa.org) issues that can halt the runtime with a SyntaxError, NameError, etc. Most other flake8 issues are merely "style violations" -- useful for readability but they do not effect runtime safety.
* F821: undefined name `name`
* F822: undefined name `name` in `__all__`
* F823: local variable name referenced before assignment
* E901: SyntaxError or IndentationError
* E999: SyntaxError -- failed to compile a file into an Abstract Syntax Tree

### CODE REVIEW

**Comment by cclauss** [comment]
Gotta love _coverage_!  Adding tests in a non-code file _lowers_ your score.

**Comment by sybrenstuvel** [comment]
I've merged half this branch, and added Python 3.7 support to Travis-CI (and removed Python 2.7 and 3.4 while I was at it). I also added MyPy, but not yet flake8.

**Comment by sybrenstuvel** [comment]
Added flake8 as dependency and fixed the things it reported in f2c3b4fd619a08a31fa49b11a0814e86169f4088

**Comment by cclauss** [comment]
Unsure how to resolve conflicts.  Both pyenv and poetry or just one of them?

**Comment by sybrenstuvel** [comment]
Only Poetry. I've never used pyenv. Did you mean pipenv? I kicked that out today in favour of Poetry.

The config for flask8 can be put in the `[flake8]` section of `setup.cfg`. That way invocations don't have to list specific settings like the maximum line length or allowed complexity. I think we can break the build for any flake8 error/warning. Currently it runs without any of those on my machine.

**Comment by cclauss** [comment]
Ready for review.

File: .travis.yml [hunk 1]
```diff
@@ -1,29 +1,19 @@
+dist: bionic
 language: python
-dist: bionic   # required for Python >= 3.7
```
Comment (sybrenstuvel) [inline .travis.yml]: Why remove the comment and move the line? I added the comment for a reason.

File: .travis.yml [hunk 2]
```diff
@@ -1,29 +1,19 @@
...
 cache: pip
-
-# Environment changes have to be manually synced with 'tox.ini'.
-# See: https://github.com/travis-ci/travis-ci/issues/3024
```
Comment (sybrenstuvel) [inline .travis.yml]: Why is this comment removed? Again, it was there for a reason.

File: .travis.yml [hunk 3]
```diff
@@ -1,29 +1,19 @@
...
+    - python: "3.5"
       dist: xenial  # Bionic has no Python 3.5
-
+    - python: "3.6"
+    - python: "3.7"
     # Disabled, see https://github.com/sybrenstuvel/python-rsa/issues/131
     #- python: pypy3.5
     #  dist: xenial  # Bionic has no Python 3.5
-
+before_install: pip install --upgrade pip
```
Comment (sybrenstuvel) [inline .travis.yml]: Why is this necessary?

File: .travis.yml [hunk 4]
```diff
@@ -1,29 +1,19 @@
...
   - pip install poetry
   - poetry install -v
-
-script:
-  - poetry run py.test tests/
-
-after_success:
-  - poetry run coveralls
+before_script: flake8 .
+script: poetry run py.test tests/
```
Comment (sybrenstuvel) [inline .travis.yml]: The `script` and `after_success` sections haven't changed, you just moved it to one line. Please only include actual changes in the commit, and leave cleanups for another PR. This only makes the diff harder to read.

File: .travis.yml [hunk 5]
```diff
@@ -1,29 +1,19 @@
...
 install:
   - pip install poetry
   - poetry install -v
-
-script:
-  - poetry run py.test tests/
-
-after_success:
-  - poetry run coveralls
+before_script: flake8 .
```
Comment (sybrenstuvel) [inline .travis.yml]: Is flake8 installed by default? Personally I'd rather run it with `poetry run flake8 .`, so that it's running the same version in Travis-CI as you'd get when running locally. Makes things a bit easier to control.

File: setup.cfg [hunk 6]
```diff
@@ -7,9 +7,8 @@ license_file = LICENSE
 [flake8]
 max-line-length = 100
 max-complexity = 10
-
-[pep8]
```
Comment (sybrenstuvel) [inline setup.cfg]: Do other tools that inspect the `[pep8]` section all have a fallback to the `[flake8]` section?

File: .travis.yml (hunk 1 above)
Comment (cclauss) [reply .travis.yml]: That comment was important when __trusty__ was the default distro on Travis CI because trusty did not support Python 3.7.

But now that xenial is the default distro, the comment is no longer needed because both xenial and bionic fully support Python 3.7  https://blog.travis-ci.com/2019-04-15-xenial-default-build-environment

File: .travis.yml (hunk 2 above)
Comment (cclauss) [reply .travis.yml]: That issue was closed in 2015.

File: .travis.yml (hunk 3 above)
Comment (cclauss) [reply .travis.yml]: Because Travis provides an older version of pip and pip is going thru rapid innovation these daze.  There is hope that pip will add a real dependency resolver in the near term.

File: .travis.yml (hunk 4 above)
Comment (cclauss) [reply .travis.yml]: It is more important that the file is concise and easier to read than the diff.  The diff will be read once, but the file hopefully many times.

File: .travis.yml (hunk 5 above)
Comment (cclauss) [reply .travis.yml]: poetry is installing flake8.

If poetry is also going to run flake8 then this line can be removed.  I am not a user of poetry (yet).

File: setup.cfg (hunk 6 above)
Comment (cclauss) [reply setup.cfg]: "pep8" was changed to "pycodestyle" several years ago at the BDFL's request.  flake8 is a superset of pycodestyle so it does everything pycodestyle did and much more besides.

File: .travis.yml (hunk 3 above)
Comment (sybrenstuvel) [reply .travis.yml]: Would be good to add a comment to this effect, including which version of Pip is now installed by default.

File: .travis.yml (hunk 4 above)
Comment (sybrenstuvel) [reply .travis.yml]: I disagree with that. Cleanups should go in a separate commit.

File: .travis.yml (hunk 5 above)
Comment (sybrenstuvel) [reply .travis.yml]: But without `poetry run` in front of it, it's not going to magically run from the Poetry-managed virtualenv.

File: setup.cfg (hunk 6 above)
Comment (sybrenstuvel) [reply setup.cfg]: Even then, such cleanups should go in a different commit. Getting flake8 to run in CI has little to do with removing unused bits of config.

File: .travis.yml (hunk 2 above)
Comment (sybrenstuvel) [reply .travis.yml]: That's irrelevant. Both files still list the same environments, and should still be kept in sync. If that's no longer necessary, only one of the files should have the list of environments, not both. Either properly handle this (in a separate commit, as it has nothing to do with getting flake8 to run in CI) or don't touch it.

File: .travis.yml (hunk 1 above)
Comment (sybrenstuvel) [reply .travis.yml]: Reasoning like this should be part of the commit message.
